### DISCUSSION TITLE
Add additional context to HelmRelease events

### DISCUSSION THREAD

**Discussion by stefansedich** [fluxcd/flux2#4718] [Proposals]
Today the events for HelmReleases do not really contain sufficient information for our monitoring needs as the event only contains the chart version, this has come up in a number of issues over the past few years and users have asked for a way to expand the metadata that these events contain.

https://github.com/fluxcd/helm-controller/issues/256
https://github.com/fluxcd/helm-controller/issues/412
https://github.com/fluxcd/helm-controller/issues/798

And this issue has been tackled by a number of abandoned PRs.

https://github.com/fluxcd/helm-controller/pull/514
https://github.com/fluxcd/helm-controller/pull/682

In our case even if we were to push `appVersion` into the event due to the way we use a shared chart for multiple application releases we really need the value of `image.tag` in the alert to make it useful for our needs.

For tools like DataDog we are able to configure them to extract release values as tags and add them to events, for example the following configuration for the [Helm check](https://docs.datadoghq.com/integrations/helm/) will add the `image.tag` release value as a tag called `helm_image_tag` on events.

```yaml
helmCheck:
  ...
  valuesAsTags:
    image.tag: helm_image_tag
```

Below are a few initial ideas I had on how to solve this problem based on our needs and previous attempts to solve this.

1. Add support for configuring the helm-controller to add additional metadata to all events based on the DataDog example above (7/10 would work but a little 🤮)
2. Provide the ability to template release values directly to alert metadata 9/10 would be nice but would require the helm-controller to add all release values to the event which I am not sure of feasibility at this stage)
3. Add the release values JSON diff to the event and then render this in the alert (8/10 would be a good start for our need :hattip: to @stefanprodan for this idea)

Would love to hear additional thoughts/ideas from maintainers so that I can work on a PR to get this issue closed off and improve the HelmRelease alert story.

**Comment by ebachle** [comment]
I don't have answers on this yet, but I'm working towards getting some at the moment.  

I had set up the fluxcd cluster check integration a few days ago, but we were still running v2.1.x and then realized upon updating to v2.2.x that the gotk_reconcile_condition that I was planning on using for monitoring the HelmRelease object was deprecated and removed.  So events might be one way I go at this.  For the moment, I'm probably going to attempt to get metrics working again with `kube-state-metrics` and seeing if I can use that to analyze the state of my cluster.

One reason I'm leaning in that direction as opposed to events currently is that on periodic refresh, I've noticed that I'd get alerted more heavily for flakey communication with HelmRepositories, GitRepositories, etc.  That's probably just a consequence of the internet being a series of tubes, but I can probably do more to smooth that long-term using metrics based alerting.

But overall at this point, I just wanted to say that I hear you on the desire for more observability into the HelmRelease lifecycle.  I'm going to work on it and report back what I find works for me.

**Reply by stefanprodan** [reply]
Flakey communication with external sources are not reflected in events emitted by helm-controller for HelmRelease upgrade operations. But a flakey internet connection can certainly result in events not showing up on Slack or any of the Flux alert providers that are hosted outside the cluster.

**Reply by ebachle** [reply]
Gotcha, with v2.1 I was seeing it emerge as HelmChart not ready, but it sounds like the v2.2 rewrite takes care of that?  Either that or the events emitted are just simply not affected by that?

**Reply by stefanprodan** [reply]
HelmChart not ready error events will not contain any chart metadata, so they are not relevant to the current discussion. If you don't care to know when charts fail to download you can drop these events. See the docs here: https://fluxcd.io/flux/components/notification/alerts/#event-exclusion

**Reply by stefanprodan** [reply]
If the Flux metrics exposed by kube-state-metrics are not sufficient, you could use https://github.com/sstarcher/helm-exporter

**Comment by stefanprodan** [comment]
## [proposal] Diff reporting for user-defined values

When helm-controller detects a change in a `HelmRelease` object, it could [calculate a diff](https://github.com/fluxcd/helm-controller/blob/main/internal/diff/unstructured.go) between the values stored in the Helm storage and `.spec.values`. 

The diff output must be filtered by [`jsondiff.DiffTypeUpdate`](https://github.com/fluxcd/helm-controller/blob/main/internal/diff/summarize.go#L33) to include only the values fields specified in `.spec.values` that differ from the Helm storage, thus avoiding leaking sensitive informations passed with `.spec.valuesFrom`.

The resulting diff report would be appended to the event message issued for Helm upgrades. For example, when image automation patches an image the output would be:

```
Helm upgrade succeeded for release default/podinfo.v2 with chart podinfo@6.6.1+0cc9a8446c95

Release values diff:

.spec.values.image.tag
-6.6.0
+6.6.1

.spec.values.image.repository
-ghcr.io/stefanprodan/podinfo
+docker.io/stefanprodan/podinfo
```

This feature would be opt-in and enabled with a feature gate e.g. `--feature-gates=EmitValuesDiffEvents=true` .

> [!IMPORTANT]
> This proposal is for reporting the diff between user-defined values specified in the HelmRelease object. Changes in the default values embedded in the chart are out of scope.

### Drawbacks

- The diff report includes only changes in `.spec.values` to avoid leaking secrets passed as values with `.spec.valuesFrom`. We could include values mapped from ConfigMaps and exclude only Secrets, but then the diff could be huge.
- The diff could leak sensitive informations even if we only include `.spec.values`, for example the AWS account ID that's part of the `image.repository` string.

**Reply by gaalw** [reply]
> thus avoiding leaking sensitive informations passed with .spec.valuesFrom.

disagree. I am passing the non-sensitive information with `valuesFrom` as well. It is very convenient - to put a `values.yaml` directly inside FluxCD repo and convert it to `ConfigMap` (or `Secret`) with means of `kustomize` and then - use `valuesFrom`.
Yes, in that case I am losing the option to patch values with kinda of image updater controller, but it is o.k.

**Reply by gaalw** [reply]
it leads to idea that one needs to put another block of configuration where he/she could set up which settings from `values.yaml` are sensitive or not. Creating additional complexity to the development and support.

**Reply by gaalw** [reply]
even more. Let's say that the helm chart was changed to new version, it brings new embedded keys in `values.yaml` and/or changing the some defaults. 

> it could [calculate a diff](https://github.com/fluxcd/helm-controller/blob/main/internal/diff/unstructured.go) between the values stored in the Helm storage and .spec.values.

should it then calculate the diff not between the Helm storage (I am not sure that we are talking about the same thing) and `.spec.values` of HR, but rather - the diff between the OLD release and rendered values of new ones? Also how would diff be calculated if I am changing chart name completely and/or release name (it's still possible)?

**Reply by stefanprodan** [reply]
So I guess you are in favour of

> We could include values mapped from ConfigMaps and exclude only Secrets

**Reply by stefanprodan** [reply]
> Also how would diff be calculated if I am changing chart name completely and/or release name (it's still possible)?

Changes to chart name or release name will result in a uninstall/install, thus no diff event will be issued since it's not an upgrade.

**Reply by stefanprodan** [reply]
@gaalw I have updated the proposal to make it clear that this is about reporting diff in user-defined values in the Flux object, diffing values embedded in the chart are out of scope.

**Reply by gecube** [reply]
> We could include values mapped from ConfigMaps and exclude only Secrets

but in the same time it means that I could not put the whole `values.yaml` to the `Secret` (and get the proper `diff`) and need to make a micromanagement and split the non-secret and secret parts of the `values.yaml`, what is not very convenient and extra work.

**Reply by gecube** [reply]
> Changes to chart name or release name will result in a uninstall/install, thus no diff event will be issued since it's not an upgrade.

Let it be so, but anyway it would be nice to have a diff even in that case. To be honest - I discourage to do so (change release name or completely change the chart name) in favour of just to remove HR and create a new one.

**Reply by stefanprodan** [reply]
> but in the same time it means that I could not put the whole values.yaml to the Secret

Yes, diffing values specified in Secrets is out of scope for my proposal as it would result in Flux CVEs.

**Reply by gecube** [reply]
> Changes in the default values embedded in the chart are out of scope.

very pity :-( It means that diff still be non-complete and the question arises could it deceive the operator of the cluster under some circumstances - i.e. we are expecting some thing one and getting something different.

**Reply by gecube** [reply]
I am not sure - but could it be more useful to get the diff not on the 'values.yaml' level, but more - show the diff in target rendered manifests? Why? because changes in `values.yaml` do not count on helpers and other functional stuff inside of Helm engine.

**Reply by stefanprodan** [reply]
We do have this diff functionality in the drift detection, but we only report the name of the fields that have drifted, instead of the actual values. The reason being that a full diff can't fit into etcd, for example the chart can contain Grafana dashboards, each close to 1Mb while the event limit is under 1Mb.

**Reply by stefansedich** [reply]
UX wise for my team the diff is not going to be as simple to grok as a single "verion" metadata field (especially if a number of superflous values change) in the alert so that would be the main drawback with this approach for us.

I could also see some values leaked that we maybe don't want to (thankfully we never put anything sensitive in values), however given a number of charts in the wild have things like "tokens" in values this could be wide-spread for other users not following best practices.

**Reply by gecube** [reply]
> The reason being that a full diff can't fit into etcd, for example the chart can contain Grafana dashboards, each close to 1Mb while the event limit is under 1Mb.

diff could be emitted just once, not to be stored in etcd as any of fields for HR... does it make sense?

**Reply by stefanprodan** [reply]
Kubernetes events are stored in etcd

**Reply by gecube** [reply]
@stefanprodan thanks, I meant "events" not in terms of k8s event, but like emitting something via webhook or similar approach to the external system. Anyway, does not matter.

**Reply by stefanprodan** [reply]
Flux events are Kubernetes events that happen to also be forwarded to external systems if you configure an alert.

**Reply by stefanprodan** [reply]
> UX wise for my team the diff is not going to be as simple to grok as a single "version" metadata field

For this I'm considering adding the `appVersion` from `Chart.yaml` to the helm-controller events. But this is a different proposal unrelated to the current one.

**Reply by stefansedich** [reply]
@stefanprodan can we perhaps provide a way to override`appVersion` with `image.tag`?

Couple reasons here:

1. it fulfills our needs where we use a shared chart that does not have `appVersion` set.
2. I see many charts in the wild that we are running today within our clusters that do not even set an `appVersion`, as mentioned we have solved this for the DataDog Helm check by mapping `values.image.tag` to an `helm_image_tag` tag within DataDog which covers the cases where `appVersion` is never set.

**Comment by stefanprodan** [comment]
In Flux v2.3, the chart app version was added to the `.status.history` snapshot and is now set as the `app-version` annotation to all the issued events. Also the app version is exposed under the `chart_app_version` label on the `gotk_resource_info` metric.

 xref: https://github.com/fluxcd/helm-controller/pull/968

**Comment by benoittgt** [comment]
Following this discussion. We would love to have more info about the source trigger of the reconciliation. For example a git sha when 
it's because of a git change. Something similar to the annotations added. When we can get it, display it.

We have many release with Flux and sometimes it's hard to know the origin. Before Flux we were tracking who trigger the helm release command, and we had custom reporting tool to inform about the progress of deployment. This is sadly something we are missing with Flux.

Meanwhile, thanks a lot for the annotations @stefanprodan . 🙌
